editoast/src/views/v2/path/path_item_cache.rs

```diff
@@ -0,0 +1,149 @@
+    pub async fn load(
+        conn: &mut DbConnection,
+        infra_id: i64,
+        path_items: &[&PathItemLocation],
+    ) -> Result<PathItemCache> {
+        let (trigrams, ops_uic, ops_id) = collect_path_item_ids(path_items);
+        let (uic_to_ops, trigram_to_ops, ids_to_ops) =
+            retrieve_op_from_locations(conn, infra_id, &ops_uic, &trigrams, &ops_id).await?;
+
+        let tracks = path_items.iter().filter_map(|item| match item {
+            PathItemLocation::TrackOffset(TrackOffset { track, .. }) => {
+                Some((infra_id, track.0.clone()))
+            }
+            _ => None,
+        });
+        let tracks = ids_to_ops
+            .values()
+            .chain(trigram_to_ops.values().flatten())
+            .chain(uic_to_ops.values().flatten())
+            .flat_map(|op| &op.parts)
+            .map(|part| (infra_id, part.track.0.clone()))
+            .chain(tracks);
+        let valid_tracks = TrackSectionModel::retrieve_batch_unchecked::<_, Vec<_>>(conn, tracks)
+            .await?
+            .into_iter()
+            .map(|track| track.obj_id)
+            .collect();
+
+        Ok(PathItemCache {
+            uic_to_ops,
+            trigram_to_ops,
+            ids_to_ops,
+            valid_tracks,
+        })
+    }
+
+    /// Get the operational points associated with an identifier
+    pub fn get_id(&self, id: &String) -> Option<&OperationalPointModel> {
+        self.ids_to_ops.get(id)
+    }
+
+    /// Get the operational points associated with a trigram
+    pub fn get_trigram(&self, trigram: &String) -> Option<&Vec<OperationalPointModel>> {
+        self.trigram_to_ops.get(trigram)
+    }
+
+    /// Get the operational points associated with a UIC code
+    pub fn get_uic(&self, uic: i64) -> Option<&Vec<OperationalPointModel>> {
+        self.uic_to_ops.get(&uic)
+    }
+
+    /// Check if a track exists
+    pub fn is_track_valid(&self, track: &String) -> bool {
+        self.valid_tracks.contains(track)
+    }
+}
+
+/// Collect the ids of the operational points from the path items
+fn collect_path_item_ids(path_items: &[&PathItemLocation]) -> (Vec<String>, Vec<i64>, Vec<String>) {
+    let mut trigrams: Vec<String> = Vec::new();
+    let mut ops_uic: Vec<i64> = Vec::new();
+    let mut ops_id: Vec<String> = Vec::new();
+
+    for item in path_items {
+        match item {
+            PathItemLocation::OperationalPointDescription { trigram, .. } => {
+                trigrams.push(trigram.clone().0);
+            }
+            PathItemLocation::OperationalPointUic { uic, .. } => {
+                ops_uic.push(*uic as i64);
+            }
+            PathItemLocation::OperationalPointId {
+                operational_point, ..
+            } => {
+                ops_id.push(operational_point.clone().0);
+            }
+            _ => {}
+        }
+    }
+    (trigrams, ops_uic, ops_id)
+}
+
+/// Retrieve operational points from locations
+async fn retrieve_op_from_locations(
```

Comment:
Why not make three separate functions ?

Reply:
Oh this is a copy paste from an old function, refactoring is kind of out of scope then

Reply:
Done